src/composeapp.cc

```diff
@@ -0,0 +1,142 @@
+                               Utils::jsonToCanonicalStr(*this));
+    }
+    return digest;
+  }
+
+  size_t archiveSize() const {
+    uint64_t arch_size{(*this)["layers"][0]["size"].asUInt64()};
+    if (0 == arch_size || arch_size > std::numeric_limits<size_t>::max()) {
+      throw std::runtime_error("Invalid size of App Archive is specified in App manifest: " +
+                               Utils::jsonToCanonicalStr(*this));
+    }
+
+    return arch_size;
+  }
+};
+
+// Utils::shell isn't interactive. The compose commands can take a few
+// seconds to run, so we use boost::process:system to stream it to stdout/sterr
+bool ComposeApp::cmd_streaming(const std::string& cmd) {
+  LOG_DEBUG << "Running: " << cmd;
+  return boost::process::system(cmd, boost::process::start_dir = root_) == 0;
+}
+
+bool ComposeApp::download(const std::string& app_uri) {
+  bool result = false;
+
+  try {
+    LOG_DEBUG << name_ << ": downloading App from Registry: " << app_uri;
+
+    Docker::Uri uri{Docker::Uri::parseUri(app_uri)};
+    Manifest manifest{registry_client_.getAppManifest(uri, Manifest::Format)};
+
+    const std::string archive_file_name{uri.digest.shortHash() + '.' + name_ + ArchiveExt};
+    Docker::Uri archive_uri{uri.createUri(manifest.archiveDigest())};
+
+    uint64_t available_storage;
+    if (checkAvailableStorageSpace(root_, available_storage)) {
+      // assume that an extracted files total size is up to 10x larger than the archive size
+      // 80% is a storage space watermark, we don't want to fill a volume above it
+      auto need_storage = manifest.archiveSize() * 10;
+      auto available_for_apps = static_cast<uint64_t>(available_storage * 0.8);
+      if (need_storage > available_for_apps) {
+        throw std::runtime_error("There is no sufficient storage space available to download App archive, available: " +
+                                 std::to_string(available_for_apps) + " need: " + std::to_string(need_storage));
+      }
+    } else {
+      LOG_WARNING << "Failed to get an available storage space, continuing with App archive download";
+    }
+
+    registry_client_.downloadBlob(archive_uri, root_ / archive_file_name, manifest.archiveSize());
+    extractAppArchive(archive_file_name);
+
+    result = true;
+    LOG_DEBUG << name_ << ": App has been downloaded";
+  } catch (const std::exception& exc) {
+    LOG_ERROR << name_ << ": failed to download App from Registry: " << exc.what();
+  }
+
+  return result;
+}
+
+bool ComposeApp::checkAvailableStorageSpace(const boost::filesystem::path& app_root, uint64_t& out_available_size) {
+  struct statvfs stat_buf {};
+  const int stat_res = statvfs(app_root.c_str(), &stat_buf);
+  if (stat_res < 0) {
+    LOG_WARNING << "Unable to read filesystem statistics: error code " << stat_res;
+    return false;
+  }
+  const uint64_t available_bytes = (static_cast<uint64_t>(stat_buf.f_bsize) * stat_buf.f_bavail);
+  const uint64_t reserved_bytes = 1 << 20;
```

Comment:
this is just copied from the original source. but could you add comment like `// 1MB reserved space`

Reply:
Sure, I forgot to do it when you asked me for the first time.

Reply:
done